### PR TITLE
[GTK][WPE] Regression: ot-svg font support lost in 2.46 with Skia

### PR DESCRIPTION
**Pull request by aperezdc**
#### 7bf0c3997b61bfa7e3410d5452fd23d82b7570ab
<pre>
[GTK][WPE] Regression: ot-svg font support lost in 2.46 with Skia
<a href="https://bugs.webkit.org/show_bug.cgi?id=283246">https://bugs.webkit.org/show_bug.cgi?id=283246</a>

Reviewed by Michael Catanzaro.

Make Skia use its own SVG render module in order to support OpenType
fonts with SVG glyph data. A new USE_SKIA_OPENTYPE_SVG build option
may be used to toggle this feature, which is enabled by default.

While it would be tempting to use WebKit&apos;s SVG renderer to handle this
case, in practice it gets complicated because handling such fonts will
attempt to render SVG while holding font-related locks, and using a
WebKit document to render SVG will try to enumerate fonts, resulting in
deadlocks. Moreover, WebKit&apos;s SVG rendering support is oriented towards
inclusion of SVG content into Web content, and is more heavyweight than
the (relatively small) Skia SVG renderer.

Based on a patchset originally by Hin-Tak Leung &lt;htl10@users.sourceforge.net&gt;.

* Source/ThirdParty/skia/CMakeLists.txt: Build SVG module and needed
  support files when the USE_SKIA_OPENTYPE_SVG option is enabled. This
  adds a new dependency on the Expat XML parser library. The symlinks
  to the include directories are rearranged in order to allow paths
  like &lt;skia/modulename/file.h&gt;.
* Source/WebKit/WebProcess/gtk/WebProcessMainGtk.cpp: Set the OpenType
  SVG decoder factory function to enable the OpenType-SVG support.
* Source/WebKit/WebProcess/wpe/WebProcessMainWPE.cpp: Ditto.
* Source/cmake/OptionsGTK.cmake: Add a new USE_SKIA_OPENTYPE_SVG option,
  enabled by default.
* Source/cmake/OptionsWPE.cmake: Ditto.
* Tools/gtk/manifest.txt.in: Include files from the Skia svg, skshaper,
  and skresources modules in tarballs.
* Tools/wpe/manifest.txt.in: Ditto.

Canonical link: <a href="https://commits.webkit.org/287859@main">https://commits.webkit.org/287859@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/eecbcaaa873d447d52a7283a1b337849703aaebc

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/81155 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/680 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/35098 "Built successfully") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/5/builds/85684 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [❌ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/32141 "Failed to checkout and rebase branch from PR 37847") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/83265 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/698 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/8490 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/5/builds/85684 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [❌ 🧪 win-tests](https://ews-build.webkit.org/#/builders/59/builds/32141 "Failed to checkout and rebase branch from PR 37847") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/84224 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/132/builds/447 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/73861 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/5/builds/85684 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/133/builds/344 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/64/builds/28023 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/30599 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/74134 "Built successfully and passed tests") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/71867 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/63/builds/28588 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/2/builds/87119 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/80213 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/8385 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/5947 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/2/builds/87119 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/8564 "Built successfully") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/8/builds/69696 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/2/builds/87119 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 vision-wk2~~](https://ews-build.webkit.org/#/builders/126/builds/14945 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/119/builds/13867 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/102620 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/12574 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/8346 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/13869 "Built successfully") | [  ~~🧪 jsc-armv7-tests~~](https://ews-build.webkit.org/#/builders/25/builds/24933 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/8183 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/11703 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/9991 "Built successfully") | | |

### CODE REVIEW

File: Source/ThirdParty/skia/CMakeLists.txt [hunk 1]
```diff
@@ -946,6 +946,86 @@ else ()
         Freetype::Freetype
     )
 endif ()
+
+if (USE_SKIA_OPENTYPE_SVG)
+    find_package(EXPAT REQUIRED)
```
Comment (mcatanzaro) [inline Source/ThirdParty/skia/CMakeLists.txt]:
```suggestion
if (USE_SKIA_OPENTYPE_SVG)
    find_package(EXPAT)
    if (NOT EXPAT_Found)
        message(FATAL_ERROR "Expat is required by USE_SKIA_OPENTYPE_SVG)
    endif ()
```

**Review by mcatanzaro — APPROVED** [review]
Well an extra SVG parser and also an extra XML parser is pretty sad from the perspective of our attack footprint, but I don't have any better ideas here, so I guess we need to....

**Comment by HinTak** [comment]
I wonder if there is a platform where cairo is not part of the system (I.e. linux) but linked in statically? wpe? Do you have a comparison of the size of cairo build vs skia build on any platform?

I compare the size of libwebkitgtk before and after, on Linux. The patch series adds about 300k to the 75MB library. Skia as a shared library (`is_component_build=true`, not a build option supported/recommended by upstream) is just over 7MB, while the svg module is just under 300k. Cairo as a set of shared libraries is only about 1.8MB in total (the main one 1.3MB).

**Comment by aperezdc** [comment]
> I wonder if there is a platform where cairo is not part of the system (I.e. linux) but linked in statically? wpe? Do you have a comparison of the size of cairo build vs skia build on any platform?

Cairo is not used with WPE at all, when building with `USE_SKIA=ON` (the default).

> I compare the size of libwebkitgtk before and after, on Linux. The patch series adds about 300k to the 75MB library. Skia as a shared library (`is_component_build=true`, not a build option supported/recommended by upstream) is just over 7MB, while the svg module is just under 300k. Cairo as a set of shared libraries is only about 1.8MB in total (the main one 1.3MB).

The difference between a WPE build with Cairo and another with Skia is ~5 MiB here on x86-64, using LTO and stripped binaries. This is with Skia statically built into WebKit, so “only” about 3.2 MiB more if we discount the size of Cairo and its dependencies.  We think this is a reasonable tradeoff, given that Skia provides more functionality than Cairo (like the option of using the GPU).

**Comment by HinTak** [comment]
Okay, so you don't have a static cairo build config now on any platform or in the past? Fair enough. One I found quite tiring is coping with the amount of upstream changes per release; we'll find out how much that impacts webkit in the future (and how often you want to sync up). Qt webengine - which uses a portion of chromium + blink + skia code - does so every 6 months, plus selective security related backports in the middle. Webkit probably doesn't need to care about syncing up as much, as skia alone is a small part and also isn't quite exposed to remote-exploitable issues.

**Pull request merged**